util/topsql/reporter/client.go

```diff
@@ -114,39 +120,48 @@ func (r *GRPCReportClient) sendBatchCPUTimeRecord(ctx context.Context, records [
 }
 
 // sendBatchSQLMeta sends a batch of SQL metas by stream.
-func (r *GRPCReportClient) sendBatchSQLMeta(ctx context.Context, metas []*tipb.SQLMeta) error {
-	if len(metas) == 0 {
-		return nil
-	}
+func (r *GRPCReportClient) sendBatchSQLMeta(ctx context.Context, sqlMap *sync.Map) error {
 	client := tipb.NewTopSQLAgentClient(r.conn)
 	stream, err := client.ReportSQLMeta(ctx)
 	if err != nil {
 		return err
 	}
-	for _, meta := range metas {
-		if err := stream.Send(meta); err != nil {
-			break
+	sqlMap.Range(func(key, value interface{}) bool {
+		sqlMeta := &tipb.SQLMeta{
+			SqlDigest:     []byte(key.(string)),
+			NormalizedSql: value.(string),
 		}
-	}
+		if err := stream.Send(sqlMeta); err != nil {
+			return false
+		}
+		return true
+	})
 	_, err = stream.CloseAndRecv()
 	return err
 }
 
-// sendBatchSQLMeta sends a batch of SQL metas by stream.
-func (r *GRPCReportClient) sendBatchPlanMeta(ctx context.Context, metas []*tipb.PlanMeta) error {
-	if len(metas) == 0 {
-		return nil
-	}
+// sendBatchPlanMeta sends a batch of SQL metas by stream.
+func (r *GRPCReportClient) sendBatchPlanMeta(ctx context.Context, planMap *sync.Map, decodePlan planBinaryDecodeFunc) error {
 	client := tipb.NewTopSQLAgentClient(r.conn)
 	stream, err := client.ReportPlanMeta(ctx)
 	if err != nil {
 		return err
 	}
-	for _, meta := range metas {
-		if err := stream.Send(meta); err != nil {
-			break
+	planMap.Range(func(key, value interface{}) bool {
+		planDecoded, err := decodePlan(value.(string))
+		if err != nil {
+			logutil.BgLogger().Warn("[top-sql] decode plan failed", zap.Error(err))
+			return false
```

Comment:
prefer to return `true` here. 1 plan decode error shouldn't affect others.

Reply:
Make sense. The agent should deal with the situation when SQL/plan meta is absent.